### PR TITLE
[CB] Easy optimizations for continuous batching

### PR DESCRIPTION
**Pull request by remi-or**
## Summary
This PR adds a few optimizations for continuous batching:
- removes non-needed `torch.cuda.synchronize` 
- add sorting of the inputs to maximize prefix caching hits
- sampling is done on the GPU
- removed an extranuous axis from the output_ids

## Performance 
| Attention         | Version       | Generated tokens | Duration (s) | Throughput (tok/s) |
|------------------------|----------------|--------------------------|------------------|--------------------------|
| Flash attention 3 | This PR         | 112599 | 16.73       | 6729.27           |
| Flash attention 3 | Main branch  | 111823 | 25.67       |  4355.68          |
| Flash attention 2 |  This PR        | 112822 | 24.61       | 4584.74          |
| Flash attention 2 | Main branch  | 112126 | 33.12       | 3385.46          |
| SDPA                  | This PR         | 113254  | 82.49       | 1373.00         |
| SDPA                  | Main branch  | 113725  | 170.84    | 665.67            |
* the performances on `main` were lowered by #42699 when compared to #42516 but this PR fixes this.

## Tests
No new failures for the CB tests. The only two failing tests are the same as in #42699 because of compile not working with `gemma2` which seems out of scope and acceptable.

### CODE REVIEW

**Comment by aymeric-roucher** [comment]
Hi sir just wanted to say this seems like a nice PR thanks

File: src/transformers/generation/continuous_batching/continuous_api.py [hunk 1]
```diff
@@ -1080,10 +1080,9 @@ def _inner_generation_loop(self, batch_processor: ContinuousBatchProcessor) -> N
             )
 
         self._generation_step()
-
-        if torch.cuda.is_available():
-            torch.cuda.synchronize()  # FIXME: why is this needed?
-        # Processor updates the batch after generation step is truly over
+        # NOTE: this was commented out and seemingly broke nothing. Keeping there in case it does
+        # if torch.cuda.is_available():
+        #     torch.cuda.synchronize()
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/continuous_api.py]:
```suggestion
```

File: tests/generation/test_continuous_batching.py [hunk 2]
```diff
@@ -251,7 +251,17 @@ def _test_continuous_batching_parity(
...
-            continuous_batching_output = continuous_batching_outputs[f"req_{i}"].generated_tokens
+            # Find the corresponding request in the continuous batching outputs
+            input_tokens = inputs.input_ids[i][inputs.attention_mask[i] == 1].tolist()
+            key_to_pop = None
+            for key, state in continuous_batching_outputs.items():
+                if state.prompt_ids == input_tokens:
+                    key_to_pop = key
+                    break
+            if key_to_pop is None:
+                self.fail(f"Request {i} not found in continuous batching outputs")
```
Comment (ArthurZucker) [inline tests/generation/test_continuous_batching.py:262]: this looks a bit complicated no? `continuous_batching_outputs[f"req_{i}"].input_tokens` ?

**Review by ArthurZucker — APPROVED** [review]

File: tests/generation/test_continuous_batching.py (hunk 2 above)
Comment (remi-or) [reply tests/generation/test_continuous_batching.py:262]: Used to be that, but we sort requests when prefix sharing is enabled, so it shuffles the order and breaks the correspondance.
